### ISSUE TITLE
[Feature] Consolidate HA `Services` and `Connections`

### ISSUE THREAD

**Issue by mmontes11** [mariadb-operator/mariadb-operator#203]
**Describe the solution you'd like**
Ability to perform writes in a single node via an extra Kubernetes `Service`. This applies to all the HA modes we currently support: Replication and Galera. This `Service` would be updated whenenver the primary changes. This is a common practice in MariaDB that helps preventing deadlocks in Galera.

The naming convention  for `Services` would be:
- `<mariadb-name>`: To be used for read requests. It will point to all nodes. 
- `<mariadb-name>-primary`: To be used for write requests. It will point to a single node, the primary.
- `<mariadb-name>-secondary`: To be used for read requests. It will point to all nodes, except the primary.

**Describe alternatives you've considered**
- We are already reconciling a primary `Service` for Replication named `<mariadb-name>-primary`
- In the case of Galera, it's something not supported yet, so the user has to create a `Service` manually. 

**Additional context**
We would need to consolidate and standarize a primary `Service` to perform writes, like CloudNativePG does. We can do so by creating a primary `Service` that would be updated by a controller whenever the `MariaDB`'s `status.currentPrimaryPodIndex` field changes.

This feature has been requested by our community multiple times:
- https://github.com/mariadb-operator/mariadb-operator/issues/202
- https://mariadb-operator.slack.com/archives/C056RAECH0W/p1691672443484259?thread_ts=1691564182.687799&cid=C056RAECH0W

**Comment by kvaps** [comment]
> To answer your question, here it is how are we planning to standarize Services for HA:
>
>  - `<mariadb-name>`: To be used for read requests. It will point to all nodes
>  - `primary-<mariadb-name>`: To be used for write requests. It will point to a single node.
>
> It will be tackled in the following releases!


The only wish I'd like to add is one more service:

 - `secondary-<mariadb-name>`: To be used only for read requests. It should point to all nodes, except the primary one.

**Comment by mmontes11** [comment]
Hey there @kvaps ! 

I do agree that this can be something very interesting to support, as it will allow us to address the replicas without hitting the primary for serving read requests.

Unfortunately, unlike `CloudNativePG`, we rely on the `StatefulSet` construct under the hood, which implies that all the `Pods` will have the same labels and therefore there is no way to selectively address only some of them. The only exception is the `statefulset.kubernetes.io/pod-name: mariadb-galera-0`, which is the label we rely on to configure the `primary-<mariadb-name>` primary service. See one of our `Pod` labels as an example:

```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    app.kubernetes.io/instance: mariadb-galera
    app.kubernetes.io/name: mariadb
    statefulset.kubernetes.io/pod-name: mariadb-galera-0
  name: mariadb-galera-0
```

For this to happen, we would need to manage `Pods` separately, not relying on a `StatefulSet` and label them based on their role. See one of the `CloudNativePG` `Pods` as an example:


```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    cnpg.io/cluster: cluster-example
    cnpg.io/instanceName: cluster-example-1
    cnpg.io/podRole: instance
    postgresql: cluster-example
    role: primary
  name: cluster-example-1
```

I'm open to ideas and suggestions!

**Comment by Guzzi** [comment]
just some thoughts:
using maxscale as sql router would allow to make readwrite split on a single port/service available (see https://mariadb.com/kb/en/mariadb-maxscale-6-read-write-splitting-with-mariadb-maxscale/) the advantage is, that the galeramon "talks" mariadb and knows about the node states (both for replication scenario and galera multimaster) and would take care about writing only to a single node andreading from multiple nodes itself - it doesn't need separate services tobe used from the outside. 
Integration into the operator would probably be not too difficult, as it just requires some functional db user and some config.
I think there are some license restrictions which I can't find right now in the documentation - so I am not sure if that is a good idea - a "kubernetes native" solution also has it's advantages - just wanted to mention that and maybe others can also give some experience to that. (I am currently running maxscale in front of both maria db operator controlled replica and galera and testing it)

**Comment by kvaps** [comment]
> The only exception is the statefulset.kubernetes.io/pod-name: mariadb-galera-0, which is the label we rely on to configure the <mariadb-name>-primary primary service. See one of our Pod labels as an example

I think managing of such labels for the every pod (better on the pod side) pretty common pattern. But if you do not want to manage labels relying on statefulset-controller, there is an another option to create and update endpoints for service directly.

**Comment by mmontes11** [comment]
> I think managing of such labels for the every pod (better on the pod side) pretty common pattern. But if you do not want to manage labels, relying on statefulset-controller, there is an another option to create and update endpoints for service directly.

That's a pretty neat suggestion, will follow this approach. Just updated the issue body. 🙏🏻 

> using maxscale as sql router would allow to make readwrite split on a single port/service available

Supporting MaxScale is definetly we want to support at some point, will create a separate issue for it. I think we should support both the "kubernetes native" `Services` and also have the ability to deploy and configure a MaxScale instance which will have also its own `Service` to make it addresable from inside and even outside (`LoadBalancer`) of the cluster.

**Comment by bluebike** [comment]
MaxScale's problem can be it's license.
It's BSL, which expires to GPL

https://github.com/mariadb-corporation/MaxScale/blob/23.02/LICENSE2302.TXT

**Comment by mmontes11** [comment]
This has been supported via the following PRs:
- https://github.com/mariadb-operator/mariadb-operator/pull/211
- https://github.com/mariadb-operator/mariadb-operator/pull/218

To be released in `v0.0.20`! Stay tuned! 🚀

**Comment by mmontes11** [comment]
This has been supported via the following PRs:
- https://github.com/mariadb-operator/mariadb-operator/pull/211
- https://github.com/mariadb-operator/mariadb-operator/pull/218

To be released in `v0.0.20` ! Stay tuned! 🚀
